h2o-algos/src/main/java/hex/glm/ComputationState.java

```diff
@@ -86,6 +113,64 @@ public ComputationState(Job job, GLMParameters parms, DataInfo dinfo, BetaConstr
     }
     _modelBetaInfo = bi;
   }
+  
+  // set _noReg to true if l1 and l2 regularization are both zero.
+  public void setNoReg() {
+    _noReg = (_lambda == 0);
+  }
+  
+  public void initConstraintInfo(LinearConstraints[] equalityConstraints, LinearConstraints[] lessThanEqualToConstraints,
+                                 List<String> coeffNames) {
+    boolean hasEqualityConstraints = equalityConstraints != null;
+    boolean hasLessConstraints = lessThanEqualToConstraints != null;
+    _derivativeEqual = hasEqualityConstraints ? calDerivatives(equalityConstraints, coeffNames) : null;
+    _derivativeLess = hasLessConstraints ? calDerivatives(lessThanEqualToConstraints, coeffNames) : null;
+    // contribution to hessian from ||h(beta)||^2 without C, stays constant once calculated, active status can change
+    _gramEqual = hasEqualityConstraints ? calGram(_derivativeEqual) : null;
+    _gramLess = hasLessConstraints ? calGram(_derivativeLess) : null;
+  }
+
+  /***
+   * Any time when the glm coefficient changes, the constraints values will change and active constraints can be inactive
+   * and vice versa.  In addition, the active status of the derivatives and 2nd derivatives can change as well.  The
+   * derivative and 2nd derivatives are part of the ComputationState.  It is the purpose of this method to change the
+   * active status of the constraint derivatives (transpose(lambda)*h(beta)) and the 2nd order derivatives of 
+   * (ck/2*transpose(h(beta))*h(beta)).
+   */
+  public void updateConstraintInfo(LinearConstraints[] equalityConstraints, LinearConstraints[] lessThanEqualToConstraints) {
+    if (equalityConstraints != null) {
+      updateDerivatives(_derivativeEqual, equalityConstraints);
+      update2ndDerivatives(_gramEqual, equalityConstraints);
+    }
+    if (lessThanEqualToConstraints != null) {
+      updateDerivatives(_derivativeLess, lessThanEqualToConstraints);
+      update2ndDerivatives(_gramLess, lessThanEqualToConstraints);
+    }
+  }
+  
+  public void update2ndDerivatives(ConstraintsGram[] constraintDerivatives, LinearConstraints[] constraints) {
+    int numConst = constraints.length;
+    for (int index=0; index<numConst; index++)
+      constraintDerivatives[index]._active = constraints[index]._active;
+  }
+  
+  public void updateDerivatives(ConstraintsDerivatives[] constraintDerivatives, LinearConstraints[] constraints) {
+    int numConst = constraints.length;
+    for (int index=0; index<numConst; index++) {
+       constraintDerivatives[index]._active = constraints[index]._active;
+    }
+  }
```

Comment:
These 2 functions look almost the same. Maybe having some interface that would be common for `ConstraintsGram` and `ConstraintsDerivatives`, this could be easily simplified.

Reply:
I combine them into one.